### PR TITLE
Modernize and fix CI

### LINKED ISSUE #119: MultiError deprecation warnings from trio
I'm starting to see warnings like
```
.../site-packages/trio_asyncio/_handles.py:131: 
TrioDeprecationWarning: trio.MultiError is deprecated since Trio 0.22.0;
use BaseExceptionGroup (on Python 3.11 and later) or exceptiongroup.BaseExceptionGroup 
(earlier versions) instead (https://github.com/python-trio/trio/issues/2211) 
```

In my code. Is there a way for trio_asyncio to get ahead of these so users don't see them at runtime?

python==3.8
trio==0.22.0
trio-asyncio==0.12.0
trio-util==0.7.0

### PR DESCRIPTION
**Pull request by oremanj**
Code changes:
* Port from MultiError to ExceptionGroup, thus supporting (and requiring) Trio v0.22.0+.
* Drop support for 3.6 (required by ExceptionGroup support). 
* Required by a new test in 3.12 but a good idea anyway: Route exceptions from asyncio callbacks through the default exception handler.
* Required by exception change above: Stop using StopAsyncIteration to indicate that the trio-asyncio loop should exit. Instead use a new TrioAsyncioExit exception, which derives from SystemExit and thus will pass through asyncio internals.
* Required by exception change above: Actually raise uncaught exceptions out of `open_loop()` when using an async loop, instead of just claiming to. Raising them out of `default_exception_handler` doesn't work because asyncio catches it. Instead we now create a new task in the loop nursery to raise the exception.
* Noticed because of exception change above: Fix a race condition where `add_reader` or `add_writer` on a socket object (rather than fd) could produce an uncaught exception if the reader/writer were unregistered and the socket closed inside the same reader/writer callback.
* Monkeypatch `asyncio.current_task` on 3.12+ so that it looks at the current_loop contextvar instead of just the C-level running loop cache. This is now necessary because `current_task` is now implemented in C and thus bypasses our monkeypatches by default.
* Clear Trio state on fork (required by new tests in 3.12).

Test changes:
* Test on 3.10, 3.11, 3.12-dev, and recent pypy.
* Move setup.cfg and pytest.ini into pyproject.toml. 
* Update the pytest plugin that runs stdlib asyncio tests so it doesn't use deprecated pytest features.

Closes #119, #120.

### CODE REVIEW

File: setup.py
Comment (agronholm) [inline setup.py:55]: Perhaps we could eliminate `setup.py` entirely since this project doesn't have any C extensions?

File: pyproject.toml
Comment (agronholm) [inline pyproject.toml:16]: Is this configuration actually used by a tool run by CI?

File: setup.py
Comment (agronholm) [inline setup.py:69]: Best not to specify a RC version, as this causes pip to install a pre-release when available.

File: README.rst
Comment (oremanj) [reply README.rst:30]: It's tested on Ubuntu. The test matrix is basically copied from current Trio; I don't know how to do prereleases on other platforms on GHA, and I don't think they'd add much additional coverage. We can add them once 3.12 final is out.

File: ci.sh
Comment (oremanj) [reply ci.sh:38]: Changed to `pip install .`

File: setup.py
Comment (oremanj) [reply setup.py:55]: I don't know how to do that and keep a single source of truth about the package version. In `setup.py` we load the version from `trio_asyncio/_version.py`.

File: pyproject.toml
Comment (oremanj) [reply pyproject.toml:16]: It was previously in `setup.cfg` so I figured I should move it over. CI doesn't currently do any linting or formatting except yapf. I'd rather not try to change that in this PR, it's already big enough.

File: setup.py
Comment (agronholm) [reply setup.py:55]: That can be done (instructions [here](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata)).

Comment (oremanj) [reply setup.py:55]: Good to know. I think I'll leave that for another PR though; this one is currently blocking a release and I don't want to complicate it further.

**Comment by dhirschfeld** [comment]
bump!

Is there anything preventing this being merged?

**Comment by oremanj** [comment]
> Is there anything preventing this being merged?

Just the fact that no one has approved it.

**Comment by dhirschfeld** [comment]
Since @agronholm has reviewed I'll let him approve if he's not too busy...

**Comment by agronholm** [comment]
I may be a little out of my depth here but I'll see if I can give this a proper review over the weekend.

**Review by agronholm — APPROVED** [review]
I didn't find anything wrong; just a note about Python version classifiers.

**Review by CoolCat467 — APPROVED** [review]
This looks good, but I would feel more comfortable if someone more familiar with the internals of asyncio had a look at this as well.

**Pull request merged**
